### PR TITLE
Simplify cost graphs and speedup costing performance

### PR DESCRIPTION
**Pull request by aslesarenko**
- memory leak fixed in (resetContext is doing additional cleanups)
- removed dependency on scalan-meta.jar
- removed dependancy on scala-compiler.jar
- various performance optimisations (based on profiling)
- costing speedup by using max constants instead of variables (which simplifies cost graph and hence reduce time)
- decreased timeout in SpamSpeciafication
- fixed limit check in ValueSerializer 
- MaxBoxSize set to 4 * 1024
- check MaxBoxSize and MaxPropositionBytes limits in deserialization

### CODE REVIEW

File: src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala [hunk 1]
```diff
@@ -122,10 +122,12 @@ object ErgoBoxCandidate {
     }
 
     def parseBodyWithIndexedDigests(digestsInTx: Option[Coll[TokenId]], r: SigmaByteReader): ErgoBoxCandidate = {
+      val previousPositionLimit = r.positionLimit
+      r.positionLimit = r.position + ErgoBox.MaxBoxSize
       val value = r.getULong()
       val tree = ErgoTreeSerializer.DefaultSerializer.deserializeErgoTree(r)
       val creationHeight = r.getUInt().toInt
-      val addTokensCount = r.getByte()
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala]: Looks like this breaks binary format?

File: src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala [hunk 2]
```diff
@@ -127,15 +128,18 @@ class ErgoTreeSerializer {
     deserializeErgoTree(r)
   }
 
-  def deserializeErgoTree(r: SigmaByteReader): ErgoTree = {
+  def deserializeErgoTree(r: SigmaByteReader,
+                          maxTreeSizeBytes: Int = SigmaSerializer.MaxPropositionSize): ErgoTree = {
```
Comment (catena2w) [inline src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala]: Is this really a function parameter? Let's either remove it from parameters list or remove the default value

File: src/test/scala/sigmastate/CostingSpecification.scala [hunk 3]
```diff
@@ -197,8 +197,10 @@ class CostingSpecification extends SigmaTestingData {
     cost(s"{ $rootTree.isInsertAllowed }") shouldBe (AccessRootHash + selectField)
     cost(s"{ $rootTree.isUpdateAllowed }") shouldBe (AccessRootHash + selectField)
     cost(s"{ $rootTree.isRemoveAllowed }") shouldBe (AccessRootHash + selectField)
-    cost(s"{ $rootTree.updateDigest($rootTree.digest) == $rootTree }") shouldBe (AccessRootHash + selectField + newAvlTreeCost + comparisonPerKbCost)
-    cost(s"{ $rootTree.updateOperations(1.toByte) == $rootTree }") shouldBe (AccessRootHash + newAvlTreeCost + comparisonPerKbCost + constCost)
+    cost(s"{ $rootTree.updateDigest($rootTree.digest) == $rootTree }") shouldBe
+      (AccessRootHash + selectField + newAvlTreeCost + comparisonCost /* for isConstantSize AvlTree type */)
+    cost(s"{ $rootTree.updateOperations(1.toByte) == $rootTree }") shouldBe
+      (AccessRootHash + newAvlTreeCost + comparisonCost + constCost)
```
Comment (catena2w) [inline src/test/scala/sigmastate/CostingSpecification.scala]: Let's add tests for changed costing rules, they should be quite easy now.

File: src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala (hunk 1 above)
Comment (greenhat) [reply src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala]: @catena2w I don't think so. We're writing unsigned values, but we were reading signed values. I just fixed reading.

Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala]: nope `@inline override def getUByte(): Int = getByte() & 0xFF`

File: src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala (hunk 2 above)
Comment (greenhat) [reply src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala]: @catena2w Probably not. Easing of testing does not count. Apart from that, decoupling from `SigmaSerializer` constant does not worth it alone. Removing default value will just force this constant leaking out. Let me remove the parameter.

Comment (aslesarenko) [reply src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala]: removed default value.

File: src/test/scala/sigmastate/CostingSpecification.scala (hunk 3 above)
Comment (aslesarenko) [reply src/test/scala/sigmastate/CostingSpecification.scala]: only this tests are affected and changed. Others are not affected.
More test will be added in #516

**Pull request merged**